flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -3571,7 +4207,326 @@ class ArrayExprLowering {
+                                                 const A &x, mlir::Type inTy) {
+    auto result = raiseComponent(cc, x, inTy, false);
+    if (isProjectedCopyInCopyOut()) {
+      auto optShape = Fortran::evaluate::GetShape(x);
+      assert(optShape.has_value());
+      convertFEShape(*optShape, explicitImpliedShape);
+    }
+    return result;
+  }
+
+  using RaiseRT = std::tuple<llvm::Optional<CC>, mlir::Type, bool, bool>;
+
+  RaiseRT
+  raiseToArray(const Fortran::evaluate::DataRef &x,
+               llvm::ArrayRef<const Fortran::semantics::Symbol *> ctrlSet) {
+    return std::visit(
+        Fortran::common::visitors{
+            [&](const Fortran::semantics::SymbolRef &s) -> RaiseRT {
+              if (x.Rank() > 0) {
+                auto [fopt2, ty2] = raiseRankedBase(s);
+                return {fopt2, ty2, false, true};
+              }
+              auto [fopt2, ty2] = raiseBase(s);
+              return {fopt2, ty2, false, false};
+            },
+            [&](const Fortran::evaluate::CoarrayRef &) -> RaiseRT {
+              TODO(getLoc(), "coarray reference");
+              return {llvm::None, mlir::Type{}, false, false};
+            },
+            [&](const auto &y) -> RaiseRT { return raiseToArray(y, ctrlSet); }},
+        x.u);
+  }
+  RaiseRT
+  raiseToArray(const Fortran::evaluate::Component &x,
+               llvm::ArrayRef<const Fortran::semantics::Symbol *> ctrlSet) {
+    auto [fopt, ty, inrank, ranked] = raiseToArray(x.base(), ctrlSet);
+    if (fopt.hasValue()) {
+      if (!ranked && x.Rank() > 0) {
+        auto [fopt2, ty2] = raiseRankedComponent(fopt, x, ty);
+        return {fopt2, ty2, ranked, true};
+      }
+      auto [fopt2, ty2] = raiseComponent(fopt, x, ty, ranked);
+      return {fopt2, ty2, ranked, ranked};
+    }
+    if (x.Rank() > 0) {
+      auto [fopt2, ty2] = raiseRankedBase(x);
+      return {fopt2, ty2, ranked, true};
+    }
+    return {fopt, ty, inrank, ranked};
+  }
+  RaiseRT raiseSubscript(const RaiseRT &tup,
+                         const Fortran::evaluate::ArrayRef &x) {
+    auto fopt = std::get<llvm::Optional<CC>>(tup);
+    if (fopt.hasValue()) {
+      auto ty = std::get<mlir::Type>(tup);
+      auto prevRanked = std::get<2>(tup);
+      auto ranked = std::get<3>(tup);
+      auto lambda = fopt.getValue();
+      auto loc = getLoc();
+      auto idxTy = builder.getIndexType();
+      auto &subs = x.subscript();
+      llvm::SmallVector<llvm::Optional<CC>> vectorCoor;
+      const auto implicitArguments = !prevRanked && ranked;
+      if (implicitArguments && isProjectedCopyInCopyOut()) {
+        // If the array reference has subscripts that use control variables
+        // from the explicit space, then those dimensions should not be
+        // considered as contributing to the implied part of the iteration
+        // space.
+        auto optShape = Fortran::evaluate::GetShape(x);
+        assert(optShape.has_value());
+        if (explicitImpliedShape.empty()) {
```

Comment:
Does `explicitImpliedShape` necessarily relies on `Fortran::evaluate::GetShape` to be computed. I did not survey all the places where it is filled up, but it looks like it is mostly set with the result of `convertFEShape`, which I only works on constant shapes. Does this mean the raising only work with constant shape arrays for now ?

Reply:
Yes, that is a limitation for the moment.